Zend/zend_operators.h

```diff
@@ -945,7 +945,7 @@ zend_memnistr(const char *haystack, const char *needle, size_t needle_len, const
 	const char *p_upper = NULL;
 	if (first_lower != first_upper) {
 		// If the needle length is 1 we don't need to look beyond p_lower as it is a guaranteed match
-		size_t upper_search_length = end - (needle_len == 1 && p_lower != NULL ? p_lower : haystack);
+		size_t upper_search_length = needle_len == 1 && p_lower != NULL ? p_lower - haystack : end - haystack;
```

Comment:
Indeed, the original code was meant to be this:

```c
		size_t upper_search_length = (needle_len == 1 && p_lower != NULL ? p_lower : end) - haystack;
```

But your version may be more readable.